---
output:
  html_document:
    toc: true
    theme: united
  knitrBootstrap::bootstrap_document:
    theme.chooser: TRUE
    highlight.chooser: TRUE
---

<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Introduction to derfinderPlot}
-->

Introduction to `derfinderPlot`
===============================

If you wish, you can view this vignette online [here](http://lcolladotor.github.io/derfinderPlot/).

```{r vignetteSetup, echo=FALSE, message=FALSE}
## Track time spent on making the vignette
startTime <- Sys.time()

## Bib setup
library('knitcitations')

## Load knitcitations with a clean bibliography
cleanbib()
cite_options(hyperlink = 'to.doc', citation_format = 'text', style = 'html')
# Note links won't show for now due to the following issue
# https://github.com/cboettig/knitcitations/issues/63

## Write bibliography information
write.bibtex(c(derfinderPlot = citation('derfinderPlot'), 
    knitrBootstrap = citation('knitrBootstrap'), 
    knitr = citation('knitr')[3],
    rmarkdown = citation('rmarkdown'),
    derPlot = citation('derfinderPlot'),
    derfinder = citation('derfinder'),
    ggbio = citation('ggbio'),
    brainspan = RefManageR::BibEntry(bibtype = 'Unpublished', key = 'brainspan', title = 'Atlas of the Developing Human Brain [Internet]. Funded by ARRA Awards 1RC2MH089921-01, 1RC2MH090047-01, and 1RC2MH089929-01.', author = 'BrainSpan', year = 2011, url = 'http://developinghumanbrain.org')),
    file = 'derfinderPlotRef.bib')
bib <- read.bibtex('derfinderPlotRef.bib')

## Fix some names to work with CRAN and GitHub versions
names(bib)[names(bib) == 'hester2013knitrbootstrap'] <- 'hester2014knitrbootstrap'
```


# Overview

`derfinderPlot` `r citep(bib[['colladotorres2014derfinderplot']])` is an addon package for `derfinder` `r citep(bib[['colladotorres2014derfinder']])` with functions that allow you to visualize the results.

While the functions in `derfinderPlot` assume you generated the data with `derfinder`, they can be used with other `GRanges` objects properly formatted.

The functions in `derfinderPlot` are:

* `plotCluster()` is a tailored `ggbio` `r citep(bib[['yin2012ggbio']])` plot that shows all the regions in a cluster (defined by distance). It shows the base-level coverage for each sample as well as the mean for each group. If these regions overlap any known gene, the gene and the transcript annotation is displayed.
* `plotOverview()` is another tailored `ggbio` `r citep(bib[['yin2012ggbio']])` plot showing an overview of the whole genome. This plot can be useful to observe if the regions are clustered in a subset of a chromosome. It can also be used to check whether the regions match predominantly one part of the gene structure (for example, 3' overlaps).
* `plotRegionCoverage()` is a fast plotting function using `R` base graphics that shows the base level coverage for each sample inside a specific region of the genome. If the region overlaps any known gene or intron, the information is displayed. Optionally, it can display the known transcripts. This function is most likely the easiest to use with `GRanges` objects from other packages.


# Example


As an example, we will analyze a small subset of the samples from the _BrainSpan Atlas of the Human Brain_ `r citep(bib[['brainspan']])` publicly available data. 

We first load the required packages.

```{r 'start', bootstrap.show.message=FALSE}
## Load libraries
supressPackageStartupMessages(library('derfinder'))
library('derfinderPlot')
```


## Analyze data

For this example, we created a small table with the relevant phenotype data for 12 samples: 6 from fetal samples and 6 from adult samples. We chose at random a brain region, in this case the primary auditory cortex (core) and for the example we will only look at data from chromosome 21. Other variables include the age in years, the gender and the RNA Integrity Number (RIN). The data is shown below.

```{r 'phenoData', bootstrap.show.code=FALSE, results = 'asis'}
library('xtable')

## Construct pheno table
pheno <- data.frame(
    gender = c('M', 'M', 'M', 'M', 'F', 'F', 'F', 'M', 'F', 'M', 'M', 'F'),
    lab = c('HSB114.A1C', 'HSB103.A1C', 'HSB178.A1C', 'HSB154.A1C', 'HSB150.A1C', 'HSB149.A1C', 'HSB130.A1C', 'HSB136.A1C', 'HSB126.A1C', 'HSB145.A1C', 'HSB123.A1C', 'HSB135.A1C'),
    Age = c(-0.642857142857143, -0.642857142857143, -0.571428571428571, -0.571428571428571, -0.666666666666667, -0.642857142857143, 21, 23, 30, 36, 37, 40),
    RIN = c(9.6, 9.7, 9.8, 9.9, 9.6, 9.4, 8.8, 9.2, 8.1, 8.4, 9.4, 8.5)
)
pheno$structure_acronym <- 'AIC'
pheno$structure_name <- 'primary auditory cortex (core)'
pheno$file <- paste0('http://download.alleninstitute.org/brainspan/MRF_BigWig_Gencode_v10/bigwig/', pheno$lab, '.bw')
pheno$group <- factor(ifelse(pheno$Age < 0, 'fetal', 'adult'), levels = c('fetal', 'adult'))

## Display the main information
p <- pheno[, -which(colnames(pheno) %in% c('structure_acronym', 'structure_name', 'file'))]
print(xtable(p), type = 'html')
```

Since the BigWig files are publicly available from _BrainSpan_ (see [here](http://download.alleninstitute.org/brainspan/MRF_BigWig_Gencode_v10/bigwig/)), we can extract the relevant coverage data using `derfinder::fullCoverage()`.

```{r 'getData', bootstrap.show.message=FALSE}
## Determine the files to use and fix the names
files <- pheno$file
names(files) <- gsub('.A1C', '', pheno$lab)

## Load the data
system.time(fullCov <- fullCoverage(files = files, chrs = 'chr21'))
```

Once we have the base-level coverage data for all 12 samples, we can construct the models. In this case, we want to find differences between fetal and adult samples while adjusting for gender, RIN and a proxy of the library size.

```{r 'models', bootstrap.show.message=FALSE}
## Get some idea of the library sizes
sampleDepths <- sampleDepth(collapseFullCoverage(fullCov), 1)

## Define models
models <- makeModels(sampleDepths, testvars = pheno$group, adjustvars = pheno[, c('gender', 'RIN')])
```

Next, we can find candidate differentially expressed regions (DERs) using as input the segments of the genome where at least one sample has coverage greater than 3. In this particular example, we chose a low theoretical F-statistic cutoff and used 20 permutations.

```{r 'analyze', bootstrap.show.message=FALSE}
## Filter coverage
filteredCov <- lapply(fullCov, filterData, cutoff = 3)

## Perform differential expression analysis
suppressPackageStartupMessages(library('bumphunter'))
system.time(results <- analyzeChr(chr = 'chr21', filteredCov$chr21, models, groupInfo = pheno$group, writeOutput = FALSE, cutoffFstat = 1e-02, nPermute = 20, seeds = 20140923 + seq_len(20)))

## Quick access to the results
regions <- results$regions$regions

## Annotation database to use
suppressPackageStartupMessages(library('TxDb.Hsapiens.UCSC.hg19.knownGene'))
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
```

## `plotOverview()`

Now that we have obtained the main results using `derfinder`, we can proceed to visualizing the results using `derfinderPlot`. The easiest to use of all the functions is `plotOverview()` which takes a set of regions and annotation information produced by `bumphunter::annotateNearest()`.

The plot below shows the candidate DERs colored by whether their q-value was less than 0.10 or not.


```{r 'plotOverview', bootstrap.show.message=FALSE, bootstrap.show.warning=FALSE}
## Q-values overview
plotOverview(regions = regions, annotation = results$annotation, type = 'qval')
```

The next plot shows the candidate DERs colored by the type of gene feature they are nearest too.

```{r 'plotOverview2', bootstrap.show.message=FALSE, bootstrap.show.warning=FALSE}
## Annotation overview
plotOverview(regions = regions, annotation = results$annotation, type = 'annotation')
```

In this particular example, because we are only using data from one chromosome the above plot is not as informative as in a real case scenario. However, with this plot we can quickly observe that nearly all of the candidate DERs are inside an exon.


## `plotRegionCoverage()`

The complete opposite of visualizing the candidate DERs at the genome level is to visualize them one region at a time. `plotRegionCoverage()` allows us to do this quickly for a large number of regions. 

Before using this function, we need to process more detailed information using two `derfinder` functions: `annotateRegions()` and `getRegionCoverage()` as shown below.

```{r 'regionData', bootstrap.show.message=FALSE}
## Get required information for the plots
annoRegs <- annotateRegions(regions, genomicState$fullGenome)
regionCov <- getRegionCoverage(fullCov, regions)
```

Once we have the relevant information we can proceed to plotting the first 20 regions. In this case, we will supply `plotRegionCoverage()` with the information it needs to plot transcripts overlapping these 20 regions.

```{r 'plotRegCov', bootstrap.show.message=FALSE}
## Plot top 20 regions
plotRegionCoverage(regions = regions, regionCoverage = regionCov, 
    groupInfo = pheno$group, nearestAnnotation = results$annotation, 
    annotatedRegions = annoRegs, whichRegions=1:20, txdb = txdb, scalefac = 1, ask = FALSE)
```

The base level coverage is shown in a log2 scale with any overlapping exons shown in dark blue and known introns in light blue. 


## `plotCluster()`

In this example, we noticed with the `plotRegionCoverage()` plots that most of the candidate DERs are contained in known exons. Sometimes, the signal might be low or we might have used very stringent cutoffs in the `derfinder` analysis. One way we can observe this is by plotting clusters of regions where a cluster is defined as regions within 300 bp (default option) of each other. 

To visualize the clusters, we can use `plotCluster()` which takes similar input to `plotOverview()` with the notable addition of the coverage information as well as the `idx` argument. This argument specifies which region to focus on: it will be plotted with a red bar and will determine the cluster to display.

In the plot below we observe one large candidate DER with other nearby ones that do not have a q-value less than 0.10. In a real analysis, we would probably discard this region as the coverage is fairly low.

```{r 'plotCluster', bootstrap.show.message=FALSE, bootstrap.show.warning=FALSE}
## First cluster
plotCluster(idx = 1, regions = regions, annotation = results$annotation, coverageInfo = fullCov$chr21, txdb = txdb, groupInfo = pheno$group)
```

The second cluster shows a larger number of potential DERs (again without q-values less than 0.10) in a segment of the genome where the coverage data is highly variable. This is a common occurrence with RNA-seq data and 

```{r 'plotCluster2', bootstrap.show.message=FALSE, bootstrap.show.warning=FALSE}
## Second cluster
plotCluster(idx = 2, regions = regions, annotation = results$annotation, coverageInfo = fullCov$chr21, txdb = txdb, groupInfo = pheno$group)
```

These plots show an ideogram which helps quickly identify which region of the genome we are focusing on. Then, the coverage level information for each sample is displayed in log2. Next, the coverage group means are shown in the log2 scale. The plot is completed with the potential and candidate DERs as well as any known transcripts.


# Advanced arguments

If you are interested in using the advanced arguments, use `derfinder::advancedArg()` as shown below:

```{r 'advancedArg'}
## URLs to advanced arguemtns
sapply(c('plotCluster', 'plotOverview'), derfinder::advancedArg, package = 'derfinderPlot', browse = FALSE)
## Set browse = TRUE if you want to open them in your browser
```






# Reproducibility

Code for creating the vignette

```{r createVignette, eval=FALSE, bootstrap.show.code=FALSE}
## Create the vignette
library('knitrBootstrap') 

knitrBootstrapFlag <- packageVersion('knitrBootstrap') < '1.0.0'
if(knitrBootstrapFlag) {
    ## CRAN version
    library('knitrBootstrap')
    system.time(knit_bootstrap('derfinderPlot.Rmd', chooser=c('boot', 'code'), show_code = TRUE))
    unlink('derfinderPlot.md')
} else {
    ## GitHub version
    library('rmarkdown')
    system.time(render('derfinderPlot.Rmd', 'knitrBootstrap::bootstrap_document'))
}
## Note: if you prefer the knitr version use:
# library('rmarkdown')
# system.time(render('derfinderPlot.Rmd', 'html_document'))
unlink('chr21', recursive = TRUE)

## Extract the R code
library('knitr')
knit('derfinderPlot.Rmd', tangle = TRUE)
```

Date the vignette was generated.

```{r reproducibility1, echo=FALSE, bootstrap.show.code=FALSE}
## Date the vignette was generated
Sys.time()
```

Wallclock time spent generating the vignette.

```{r reproducibility2, echo=FALSE, bootstrap.show.code=FALSE}
## Processing time in seconds
totalTime <- diff(c(startTime, Sys.time()))
round(totalTime, digits=3)
```

`R` session information.

```{r reproducibility3, echo=FALSE, bootstrap.show.code=FALSE, bootstrap.show.message=FALSE}
## Session info
library('devtools')
session_info()
```

# Bibliography

This vignette was generated using `knitrBootstrap` `r citep(bib[['hester2014knitrbootstrap']])`
with `knitr` `r citep(bib[['xie2014knitr']])` and `rmarkdown` `r citep(bib[['allaire2014rmarkdown']])` running behind the scenes.

Citations made with `knitcitations` `r citep('https://github.com/cboettig/knitcitations')`.

```{r vignetteBiblio, results='asis', echo=FALSE}
## Print bibliography
bibliography()
```
